skimage/filters/rank/core_cy.pyx

```diff
@@ -271,3 +286,293 @@ cdef void _core(void kernel(dtype_t_out*, Py_ssize_t, Py_ssize_t*, double,
+                se_w_r is NULL or se_w_c is NULL or se_w_p is NULL or
+                se_n_r is NULL or se_n_c is NULL or se_n_p is NULL or
+                se_s_r is NULL or se_s_c is NULL or se_s_p is NULL or
+                histo is NULL):
+            free(se_e_r)
+            free(se_e_c)
+            free(se_e_p)
+            free(se_w_r)
+            free(se_w_c)
+            free(se_w_p)
+            free(se_n_r)
+            free(se_n_c)
+            free(se_n_p)
+            free(se_s_r)
+            free(se_s_c)
+            free(se_s_p)
+            free(histo)
+            with gil:
+                raise MemoryError()
+
+        for i in range(n_bins):
+            histo[i] = 0
+
+        num_se_n = num_se_s = num_se_e = num_se_w = 0
+
+        for r in range(srows):
+            for c in range(scols):
+                for p in range(splanes):
+                    if t_e[p, r, c]:
+                        se_e_r[num_se_e] = r - centre_r
+                        se_e_c[num_se_e] = c - centre_c
+                        se_e_p[num_se_e] = p - centre_p
+                        num_se_e += 1
+                    if t_w[p, r, c]:
+                        se_w_r[num_se_w] = r - centre_r
+                        se_w_c[num_se_w] = c - centre_c
+                        se_w_p[num_se_w] = p - centre_p
+                        num_se_w += 1
+                    if t_n[p, r, c]:
+                        se_n_r[num_se_n] = r - centre_r
+                        se_n_c[num_se_n] = c - centre_c
+                        se_n_p[num_se_n] = p - centre_p
+                        num_se_n += 1
+                    if t_s[p, r, c]:
+                        se_s_r[num_se_s] = r - centre_r
+                        se_s_c[num_se_s] = c - centre_c
+                        se_s_p[num_se_s] = p - centre_p
+                        num_se_s += 1
+
+        for p in range(planes):
+            for i in range(n_bins):
+                histo[i] = 0
+            pop = 0
+
+            for r in range(srows):
+                for c in range(scols):
+                    for j in range(splanes):
+                        rr = r - centre_r
+                        cc = c - centre_c
+                        pp = j - centre_p + p
+
+                        # change this to accept 3D
+                        if selem[j, r, c]:
+                            if is_in_mask_3D(rows, cols, planes, rr, cc, pp,
+                                             mask_data):
+                                histogram_increment(histo, &pop,
+                                                    image[pp, rr, cc])
+
+            r = 0
+            c = 0
+            kernel(&out[p, r, c, 0], odepth, histo, pop, image[p, r, c],
+                   n_bins, mid_bin, p0, p1, s0, s1)
+
+        # main loop
+
+            r = 0
+            for even_row in range(0, rows, 2):
+
+                # ---> west to east
+                for c in range(1, cols):
```

Comment:
I don't understand why things start at 1 here?

Reply:
I see that you have applied the kernel at r=0, c=0, before you even start, above, but that only applies to the top-left pixel? What about the pixels at [p, 2, 0], [p, 4, 0], ...?

Reply:
I believe this is also copied from the 2D version. But I will look into this.

Reply:
I think I figured it out: the 0th pixel gets processed when moving N to S in the last iteration of the loop. Is that right @ryanlu41?